File: typescript/src/resilience.ts
/**
 * LLMix Resilience Module
 *
 * Circuit breaker, kill switch, singleflight deduplication, and retry with
 * exponential backoff + jitter.
 */

import { createHash } from "node:crypto";
import { existsSync, mkdirSync, renameSync, statSync } from "node:fs";
import { stat } from "node:fs/promises";
import { homedir } from "node:os";
import { join } from "node:path";

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

const DEFAULT_FAILURE_THRESHOLD = 3;
const DEFAULT_COOLDOWN_MS = 30_000;
const DEFAULT_PERMITTED_HALF_OPEN_CALLS = 10;
const DEFAULT_BASE_DELAY_MS = 1_000;
const DEFAULT_MAX_DELAY_MS = 30_000;
const DEFAULT_JITTER_MS = 1_000;
const DEFAULT_MAX_RETRY_AFTER_MS = 60_000;
const KILLSWITCH_FILENAME = "killswitch";
const KILLSWITCH_SUBDIR = "llmix";
const LEGACY_KILLSWITCH_SUBDIR = "llmix2";

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function isRetryableStatus(statusCode: number): boolean {
  return statusCode === 408 || statusCode === 429 || (statusCode >= 500 && statusCode <= 599);
}

function resolveStateDir(): string {
  const envDir = process.env["LLMIX_STATE_DIR"];
  if (envDir) return envDir;

  const xdg = process.env["XDG_STATE_HOME"];
  if (xdg) return join(xdg, KILLSWITCH_SUBDIR);

  return join(homedir(), ".local", "state", KILLSWITCH_SUBDIR);
}

function resolveLegacyStateDir(currentStateDir: string): string | null {
  const parts = currentStateDir.split(/[/\\]+/);
  if (parts.at(-1) !== KILLSWITCH_SUBDIR) {
    return null;
  }
  return join(currentStateDir, "..", LEGACY_KILLSWITCH_SUBDIR);
}

function migrateLegacyKillSwitch(currentStateDir: string): string {
  const currentPath = join(currentStateDir, KILLSWITCH_FILENAME);
  const legacyStateDir = resolveLegacyStateDir(currentStateDir);
  if (legacyStateDir === null || existsSync(currentPath)) {
    return currentPath;
  }

  const legacyPath = join(legacyStateDir, KILLSWITCH_FILENAME);
  if (!existsSync(legacyPath)) {
    return currentPath;
  }

  mkdirSync(currentStateDir, { recursive: true });
  renameSync(legacyPath, currentPath);
  console.warn(`[llmix] migrated legacy kill switch from ${legacyPath} to ${currentPath}`);
  return currentPath;
}

// ---------------------------------------------------------------------------
// Errors
// ---------------------------------------------------------------------------

export class CircuitOpenError extends Error {
  readonly provider: string;
  readonly baseUrl: string;

  constructor(provider: string, baseUrl: string) {
    super(`Circuit breaker OPEN for (${provider}, ${baseUrl})`);
    this.name = "CircuitOpenError";
    this.provider = provider;
    this.baseUrl = baseUrl;
    Object.setPrototypeOf(this, CircuitOpenError.prototype);
  }
}

export class KillSwitchActiveError extends Error {
  readonly path: string;

  constructor(path: string) {
    super(`Kill switch active: ${path} exists. All LLMix calls are blocked.`);
    this.name = "KillSwitchActiveError";
    this.path = path;
    Object.setPrototypeOf(this, KillSwitchActiveError.prototype);
  }
}

// ---------------------------------------------------------------------------
// Circuit Breaker
// ---------------------------------------------------------------------------

export enum CircuitState {
  CLOSED = "CLOSED",
  OPEN = "OPEN",
  HALF_OPEN = "HALF_OPEN",
}

export interface CircuitBreakerOptions {
  failureThreshold?: number;
  cooldownMs?: number;
  permittedHalfOpenCalls?: number;
}

export class CircuitBreaker {
  readonly provider: string;
  readonly baseUrl: string;
  readonly failureThreshold: number;
  readonly permittedHalfOpenCalls: number;
  cooldownMs: number;

  private _state = CircuitState.CLOSED;
  private _consecutiveFailures = 0;
  private _openedAt = 0;
  private _baseCooldownMs: number;
  // HALF_OPEN probe tracking (Resilience4j-style)
  private _halfOpenActive = 0;
  private _halfOpenSuccesses = 0;
  private _halfOpenFailures = 0;

  constructor(
    provider: string,
    baseUrl: string,
    options?: CircuitBreakerOptions,
  ) {
    this.provider = provider;
    this.baseUrl = baseUrl;
    this.failureThreshold = options?.failureThreshold ?? DEFAULT_FAILURE_THRESHOLD;
    this.cooldownMs = options?.cooldownMs ?? DEFAULT_COOLDOWN_MS;
    this._baseCooldownMs = this.cooldownMs;
    this.permittedHalfOpenCalls = options?.permittedHalfOpenCalls ?? DEFAULT_PERMITTED_HALF_OPEN_CALLS;
  }

  get state(): CircuitState {
    if (this._state === CircuitState.OPEN) {
      const elapsed = performance.now() - this._openedAt;
      if (elapsed >= this.cooldownMs) {
        this._state = CircuitState.HALF_OPEN;
        this._halfOpenActive = 0;
        this._halfOpenSuccesses = 0;
        this._halfOpenFailures = 0;
        return CircuitState.HALF_OPEN;
      }
    }
    return this._state;
  }

  check(): void {
    const current = this.state;
    if (current === CircuitState.CLOSED) return;

    if (current === CircuitState.HALF_OPEN) {
      if (this._halfOpenActive >= this.permittedHalfOpenCalls) {
        throw new CircuitOpenError(this.provider, this.baseUrl);
      }
      this._halfOpenActive++;
      return;
    }

    // OPEN
    throw new CircuitOpenError(this.provider, this.baseUrl);
  }

  /**
   * Evaluate HALF_OPEN results once enough probes have completed.
   *
   * Uses a fixed window (permittedHalfOpenCalls) so we wait for the full
   * sample before deciding. If a probe is lost (timeout/crash), the failure
   * path in cancelProbe() counts it as a failure so the window always completes.
   */
  private _evaluateHalfOpen(): void {
    const totalCompleted = this._halfOpenSuccesses + this._halfOpenFailures;
    if (totalCompleted < this.permittedHalfOpenCalls) {
      return; // Need more samples before deciding
    }

    if (this._halfOpenSuccesses > this._halfOpenFailures) {
      // Majority success — service recovered
      this._state = CircuitState.CLOSED;
      this._consecutiveFailures = 0;
      this.cooldownMs = this._baseCooldownMs;
    } else {
      // Majority failures — re-open with exponential backoff (cap 5 min)
      this._state = CircuitState.OPEN;
      this._openedAt = performance.now();
      this.cooldownMs = Math.min(this.cooldownMs * 2, 300_000);
    }
  }

  onSuccess(): void {
    if (this._state === CircuitState.HALF_OPEN) {
      this._halfOpenSuccesses++;
      this._evaluateHalfOpen();
      return;
    }
    // CLOSED state — reset failure counter
    this._state = CircuitState.CLOSED;
    this._consecutiveFailures = 0;
  }

  onFailure(statusCode?: number, networkError = false): void {
    let retryable = networkError;
    if (statusCode !== undefined) {
      retryable = retryable || isRetryableStatus(statusCode);
    }

    if (this._state === CircuitState.HALF_OPEN) {
      if (retryable || networkError) {
        this._halfOpenFailures++;
      } else {
        // Non-retryable (400, 404) = server is reachable, count as success
        this._halfOpenSuccesses++;
      }
      this._evaluateHalfOpen();
      return;
    }

    // Auth errors do NOT trip the breaker in CLOSED state.
    // Server is reachable → reset the consecutive failure counter.
    if (statusCode !== undefined && (statusCode === 401 || statusCode === 403)) {
      this._consecutiveFailures = 0;
      return;
    }

    if (!retryable) {
      // Non-retryable error proves server is reachable — reset counter
      this._consecutiveFailures = 0;
      return;
    }

    this._consecutiveFailures++;
    if (this._consecutiveFailures >= this.failureThreshold) {
      this._state = CircuitState.OPEN;
      this._openedAt = performance.now();
    }
  }

  /**
   * Cancel an in-flight HALF_OPEN probe without recording success or failure.
   *
   * Safety net for when onSuccess/onFailure were never called (e.g. crash).
   * No-op if the probe was already finalized — prevents double-counting when
   * onFailure() and cancelProbe() both fire for the same probe.
   */
  cancelProbe(): void {
    if (this._state !== CircuitState.HALF_OPEN) return;
    const totalFinalized = this._halfOpenSuccesses + this._halfOpenFailures;
    if (totalFinalized >= this._halfOpenActive) {
      return; // All admitted probes already reported — this is a duplicate
    }
    this._halfOpenFailures++;
    this._evaluateHalfOpen();
  }

  reset(): void {
    this._state = CircuitState.CLOSED;
    this._consecutiveFailures = 0;
    this._openedAt = 0;
    this._halfOpenActive = 0;
    this._halfOpenSuccesses = 0;
    this._halfOpenFailures = 0;
    this.cooldownMs = this._baseCooldownMs;
  }
}

// ---------------------------------------------------------------------------
// Kill Switch
// ---------------------------------------------------------------------------

export class KillSwitch {
  readonly path: string;

  constructor(stateDir?: string) {
    const base = stateDir ?? resolveStateDir();
    this.path = migrateLegacyKillSwitch(base);
  }

  check(): void {
    try {
      statSync(this.path);
      throw new KillSwitchActiveError(this.path);
    } catch (err: unknown) {
      if (err instanceof KillSwitchActiveError) throw err;
      if ((err as NodeJS.ErrnoException).code === "ENOENT") return;
      throw err; // EACCES, EPERM, etc. — don't silently swallow
    }
  }

  isActive(): boolean {
    try {
      statSync(this.path);
      return true;
    } catch (err: unknown) {
      if ((err as NodeJS.ErrnoException).code === "ENOENT") return false;
      throw err;
    }
  }

  /** Async check — avoids blocking the event loop with statSync. */
  async checkAsync(): Promise<void> {
    try {
      await stat(this.path);
      throw new KillSwitchActiveError(this.path);
    } catch (err: unknown) {
      if (err instanceof KillSwitchActiveError) throw err;
      if ((err as NodeJS.ErrnoException).code === "ENOENT") return;
      throw err;
    }
  }

  /** Async version of isActive(). */
  async isActiveAsync(): Promise<boolean> {
    try {
      await stat(this.path);
      return true;
    } catch (err: unknown) {
      if ((err as NodeJS.ErrnoException).code === "ENOENT") return false;
      throw err;
    }
  }
}

// ---------------------------------------------------------------------------
// Singleflight
// ---------------------------------------------------------------------------

interface FlightEntry<T> {
  promise: Promise<T>;
  resolve: (value: T) => void;
  reject: (reason: unknown) => void;
}

export class Singleflight {
  private _inFlight = new Map<string, FlightEntry<unknown>>();

  static makeKey(data: string): string {
    return createHash("sha256").update(data).digest("hex");
  }

  async do<T>(key: string, fn: () => Promise<T>): Promise<T> {
    const existing = this._inFlight.get(key);
    if (existing) {
      // Safety: FlightEntry is stored as <unknown> because the Map can't be
      // generic per-key.  The cast is safe because callers sharing the same key
      // always produce the same T — the singleflight contract guarantees it.
      return existing.promise as Promise<T>;
    }

    let resolve!: (value: T) => void;
    let reject!: (reason: unknown) => void;
    const promise = new Promise<T>((res, rej) => {
      resolve = res;
      reject = rej;
    });

    // Prevent unhandled rejection when no second caller awaits this promise
    promise.catch(() => {});

    const entry: FlightEntry<T> = { promise, resolve, reject };
    this._inFlight.set(key, entry as FlightEntry<unknown>);

    try {
      const result = await fn();
      resolve(result);
      return result;
    } catch (err: unknown) {
      reject(err);
      throw err;
    } finally {
      this._inFlight.delete(key);
    }
  }

  get inFlightCount(): number {
    return this._inFlight.size;
  }
}

// ---------------------------------------------------------------------------
// Retry with Exponential Backoff + Jitter
// ---------------------------------------------------------------------------

export function calculateDelay(
  attempt: number,
  options?: {
    baseMs?: number;
    maxDelayMs?: number;
    jitterMs?: number;
  },
): number {
  const baseMs = options?.baseMs ?? DEFAULT_BASE_DELAY_MS;
  const maxDelayMs = options?.maxDelayMs ?? DEFAULT_MAX_DELAY_MS;
  const jitterMs = options?.jitterMs ?? DEFAULT_JITTER_MS;

  const exponential = Math.min(2 ** attempt * baseMs, maxDelayMs);
  const jitter = Math.floor(Math.random() * (jitterMs + 1));
  return exponential + jitter;
}

export function parseRetryAfter(
  headerValue: string | undefined | null,
  maxMs = DEFAULT_MAX_RETRY_AFTER_MS,
): number | null {
  if (headerValue == null) return null;
  const trimmed = headerValue.trim();
  if (/^\d+$/.test(trimmed)) {
    const seconds = parseInt(trimmed, 10);
    return Math.min(seconds * 1000, maxMs);
  }

  // Fallback: try HTTP-date format (RFC 7231 §7.1.1.1)
  const date = new Date(headerValue);
  if (!isNaN(date.getTime())) {
    const deltaMs = date.getTime() - Date.now();
    if (deltaMs > 0) return Math.min(deltaMs, maxMs);
  }

  return null;
}

export function isRetryable(statusCode: number): boolean {
  return isRetryableStatus(statusCode);
}

export interface RetryPolicyOptions {
  maxRetries?: number | undefined;
  baseMs?: number | undefined;
  maxDelayMs?: number | undefined;
  jitterMs?: number | undefined;
  maxRetryAfterMs?: number | undefined;
}

export class RetryPolicy {
  readonly maxRetries: number;
  readonly baseMs: number;
  readonly maxDelayMs: number;
  readonly jitterMs: number;
  readonly maxRetryAfterMs: number;

  constructor(options?: RetryPolicyOptions) {
    this.maxRetries = options?.maxRetries ?? 3;
    this.baseMs = options?.baseMs ?? DEFAULT_BASE_DELAY_MS;
    this.maxDelayMs = options?.maxDelayMs ?? DEFAULT_MAX_DELAY_MS;
    this.jitterMs = options?.jitterMs ?? DEFAULT_JITTER_MS;
    this.maxRetryAfterMs = options?.maxRetryAfterMs ?? DEFAULT_MAX_RETRY_AFTER_MS;

    if (!Number.isInteger(this.maxRetries) || this.maxRetries < 0) {
      throw new RangeError(`maxRetries must be a non-negative integer, got ${this.maxRetries}`);
    }
    if (!Number.isFinite(this.baseMs) || this.baseMs < 0) {
      throw new RangeError(`baseMs must be a non-negative finite number, got ${this.baseMs}`);
    }
    if (!Number.isFinite(this.maxDelayMs) || this.maxDelayMs < 0) {
      throw new RangeError(`maxDelayMs must be a non-negative finite number, got ${this.maxDelayMs}`);
    }
    if (!Number.isFinite(this.jitterMs) || this.jitterMs < 0) {
      throw new RangeError(`jitterMs must be a non-negative finite number, got ${this.jitterMs}`);
    }
    if (!Number.isFinite(this.maxRetryAfterMs) || this.maxRetryAfterMs < 0) {
      throw new RangeError(`maxRetryAfterMs must be a non-negative finite number, got ${this.maxRetryAfterMs}`);
    }
  }

  getDelayMs(attempt: number, retryAfterHeader?: string | null): number {
    const retryAfter = parseRetryAfter(retryAfterHeader, this.maxRetryAfterMs);
    if (retryAfter !== null) return retryAfter;

    return calculateDelay(attempt, {
      baseMs: this.baseMs,
      maxDelayMs: this.maxDelayMs,
      jitterMs: this.jitterMs,
    });
  }

  async execute<T>(
    fn: () => Promise<T>,
    isRetryableFn?: (err: unknown) => boolean,
  ): Promise<T> {
    let lastError: unknown;
    for (let attempt = 0; attempt <= this.maxRetries; attempt++) {
      try {
        return await fn();
      } catch (err: unknown) {
        lastError = err;
        if (attempt >= this.maxRetries) throw err;
        if (isRetryableFn && !isRetryableFn(err)) throw err;
        const retryAfterHeader = (err as { headers?: Record<string, string> }).headers?.["retry-after"] ?? null;
        const delayMs = this.getDelayMs(attempt, retryAfterHeader);
        await new Promise((r) => setTimeout(r, delayMs));
      }
    }
    throw lastError;
  }
}

// ---------------------------------------------------------------------------
// Cross-Process File Lock (opt-in via LLM_GLOBAL_CONCURRENCY)
//
// Uses `proper-lockfile` npm package when available.
// No-op when LLM_GLOBAL_CONCURRENCY env var is not set.
// ---------------------------------------------------------------------------

export interface FileLockLike {
  acquire(): Promise<void>;
  release(): Promise<void>;
  readonly enabled: boolean;
}

export async function createFileLock(
  lockPath?: string,
): Promise<FileLockLike> {
  const concurrency = process.env["LLM_GLOBAL_CONCURRENCY"];
  const enabled = concurrency !== undefined && concurrency.trim() !== "";

  if (enabled && (!/^\d+$/.test(concurrency.trim()) || parseInt(concurrency.trim(), 10) < 1)) {
    throw new RangeError(
      `LLM_GLOBAL_CONCURRENCY must be a positive integer, got "${concurrency}"`,
    );
  }

  if (!enabled) {
    return {
      enabled: false,
      async acquire() { /* no-op */ },
      async release() { /* no-op */ },
    };
  }

  const base = resolveStateDir();
  const resolvedPath = lockPath ?? join(base, "llmix.lock");

  // Dynamic import so proper-lockfile is only loaded when needed
  let lockfileMod: { lock: (path: string, options?: Record<string, unknown>) => Promise<() => Promise<void>> };
  try {
    lockfileMod = await import("proper-lockfile") as typeof lockfileMod;
  } catch (cause: unknown) {
    throw new Error(
      "LLMix: LLM_GLOBAL_CONCURRENCY is set but 'proper-lockfile' could not be loaded. " +
      "Install with: bun add proper-lockfile",
      { cause },
    );
  }

  let releaseFn: (() => Promise<void>) | null = null;

  return {
    enabled: true,
    async acquire() {
      // Ensure parent directory exists
      const { mkdir, writeFile } = await import("node:fs/promises");
      const { dirname } = await import("node:path");
      await mkdir(dirname(resolvedPath), { recursive: true });
      // proper-lockfile needs the file to exist
      try {
        await writeFile(resolvedPath, "", { flag: "wx" });
      } catch (err: unknown) {
        if ((err as NodeJS.ErrnoException).code !== "EEXIST") throw err;
      }
      releaseFn = await lockfileMod.lock(resolvedPath, { retries: { retries: 10, minTimeout: 100 } });
    },
    async release() {
      if (releaseFn) {
        await releaseFn();
        releaseFn = null;
      }
    },
  };
}

// ---------------------------------------------------------------------------
// Re-export helpers
// ---------------------------------------------------------------------------

export { resolveStateDir };

File: python/llmix/resilience.py
"""
LLMix Resilience Module

Circuit breaker, kill switch, singleflight deduplication, and retry with
exponential backoff + jitter. All four primitives live in this single module
to keep the dependency graph flat.
"""

import asyncio
import enum
import hashlib
import os
import random
import time
from collections.abc import Awaitable, Callable
from pathlib import Path
from typing import Any, TypeVar

try:
    import fcntl as _fcntl
except ImportError:
    _fcntl = None  # Windows

T = TypeVar("T")

# ---------------------------------------------------------------------------
# Constants
# ---------------------------------------------------------------------------

_DEFAULT_FAILURE_THRESHOLD = 3
_DEFAULT_COOLDOWN_SECONDS = 30.0
_DEFAULT_MAX_DELAY_MS = 30_000
_DEFAULT_BASE_DELAY_MS = 1_000
_DEFAULT_JITTER_MS = 1_000
_DEFAULT_MAX_RETRY_AFTER_MS = 60_000
_KILLSWITCH_FILENAME = "killswitch"
_KILLSWITCH_SUBDIR = "llmix2"


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------


def _is_retryable_status(status_code: int) -> bool:
    """Return True for 408, 429, and 5xx status codes."""
    return status_code == 408 or status_code == 429 or 500 <= status_code <= 599


def _resolve_state_dir() -> Path:
    """Resolve the LLMix state directory.

    Priority:
    1. LLMIX_STATE_DIR env var
    2. XDG_STATE_HOME/llmix2
    3. ~/.local/state/llmix2
    """
    env_dir = os.environ.get("LLMIX_STATE_DIR")
    if env_dir:
        return Path(env_dir)

    xdg = os.environ.get("XDG_STATE_HOME")
    if xdg:
        return Path(xdg) / _KILLSWITCH_SUBDIR

    return Path.home() / ".local" / "state" / _KILLSWITCH_SUBDIR


# ---------------------------------------------------------------------------
# Errors
# ---------------------------------------------------------------------------


class CircuitOpenError(Exception):
    """Raised when the circuit breaker is OPEN and rejecting calls."""

    def __init__(self, provider: str, base_url: str) -> None:
        self.provider = provider
        self.base_url = base_url
        super().__init__(f"Circuit breaker OPEN for ({provider}, {base_url})")


class KillSwitchActiveError(Exception):
    """Raised when the kill switch file is present."""

    def __init__(self, path: Path) -> None:
        self.path = path
        super().__init__(f"Kill switch active: {path} exists. All LLMix calls are blocked.")


# ---------------------------------------------------------------------------
# Circuit Breaker
# ---------------------------------------------------------------------------


class CircuitState(enum.Enum):
    CLOSED = "CLOSED"
    OPEN = "OPEN"
    HALF_OPEN = "HALF_OPEN"


_DEFAULT_PERMITTED_HALF_OPEN_CALLS = 10


class CircuitBreaker:
    """Per-(provider, baseUrl) circuit breaker (Resilience4j-style).

    State machine:
      CLOSED    -- N consecutive retryable failures --> OPEN
      OPEN      -- cooldown expires --> HALF_OPEN
      HALF_OPEN -- allows up to `permitted_half_open_calls` concurrent probes
                   if probe success rate >= 50% --> CLOSED
                   if probe failure rate > 50%  --> OPEN (with doubled cooldown)
    """

    def __init__(
        self,
        provider: str,
        base_url: str,
        failure_threshold: int = _DEFAULT_FAILURE_THRESHOLD,
        cooldown_seconds: float = _DEFAULT_COOLDOWN_SECONDS,
        permitted_half_open_calls: int = _DEFAULT_PERMITTED_HALF_OPEN_CALLS,
    ) -> None:
        self.provider = provider
        self.base_url = base_url
        self.failure_threshold = failure_threshold
        self.cooldown_seconds = cooldown_seconds
        self.permitted_half_open_calls = permitted_half_open_calls
        self._base_cooldown = cooldown_seconds

        self._state = CircuitState.CLOSED
        self._consecutive_failures = 0
        self._opened_at: float = 0.0
        # HALF_OPEN probe tracking (Resilience4j-style)
        self._half_open_active: int = 0
        self._half_open_successes: int = 0
        self._half_open_failures: int = 0

    @property
    def state(self) -> CircuitState:
        """Current state, accounting for cooldown transitions."""
        if self._state == CircuitState.OPEN:
            elapsed = time.monotonic() - self._opened_at
            if elapsed >= self.cooldown_seconds:
                self._state = CircuitState.HALF_OPEN
                self._half_open_active = 0
                self._half_open_successes = 0
                self._half_open_failures = 0
                return CircuitState.HALF_OPEN
        return self._state

    def check(self) -> None:
        """Check whether a request is allowed.

        Raises CircuitOpenError if the breaker is OPEN or HALF_OPEN slots full.
        In HALF_OPEN, allows up to `permitted_half_open_calls` concurrent probes.
        """
        current = self.state
        if current == CircuitState.CLOSED:
            return
        if current == CircuitState.HALF_OPEN:
            if self._half_open_active >= self.permitted_half_open_calls:
                raise CircuitOpenError(self.provider, self.base_url)
            self._half_open_active += 1
            return
        # OPEN
        raise CircuitOpenError(self.provider, self.base_url)

    def _evaluate_half_open(self) -> None:
        """Evaluate HALF_OPEN results once enough probes have completed.

        Uses a fixed window (permitted_half_open_calls) — not the active count —
        so we wait for the full sample before deciding. If a probe is lost
        (timeout/crash), the failure path in cancel_probe() counts it as a
        failure so the window always completes.
        """
        total_completed = self._half_open_successes + self._half_open_failures
        if total_completed < self.permitted_half_open_calls:
            return  # Need more samples before deciding

        if self._half_open_successes > self._half_open_failures:
            # Majority success — service recovered
            self._state = CircuitState.CLOSED
            self._consecutive_failures = 0
            self.cooldown_seconds = self._base_cooldown
        else:
            # Majority failures — re-open with exponential backoff (cap 5 min)
            self._state = CircuitState.OPEN
            self._opened_at = time.monotonic()
            self.cooldown_seconds = min(self.cooldown_seconds * 2, 300.0)

    def on_success(self) -> None:
        """Record a successful request."""
        if self._state == CircuitState.HALF_OPEN:
            self._half_open_successes += 1
            self._evaluate_half_open()
            return
        # CLOSED state — reset failure counter
        self._state = CircuitState.CLOSED
        self._consecutive_failures = 0

    def on_failure(self, status_code: int | None = None, *, network_error: bool = False) -> None:
        """Record a failed request.

        Only retryable errors (408, 429, 5xx, network errors) increment the counter.
        Auth errors (401, 403) are ignored in CLOSED state but still finalize
        the HALF_OPEN probe.
        """
        retryable = network_error
        if status_code is not None:
            retryable = retryable or _is_retryable_status(status_code)

        if self._state == CircuitState.HALF_OPEN:
            if retryable or network_error:
                self._half_open_failures += 1
            else:
                # Non-retryable (400, 404) = server is reachable, count as success
                self._half_open_successes += 1
            self._evaluate_half_open()
            return

        # Auth errors do NOT trip the breaker in CLOSED state.
        if status_code is not None and status_code in (401, 403):
            self._consecutive_failures = 0
            return

        if not retryable:
            self._consecutive_failures = 0
            return

        self._consecutive_failures += 1
        if self._consecutive_failures >= self.failure_threshold:
            self._state = CircuitState.OPEN
            self._opened_at = time.monotonic()

    def cancel_probe(self) -> None:
        """Cancel an in-flight HALF_OPEN probe without recording success or failure.

        Safety net for when on_success/on_failure were never called (e.g. crash).
        No-op if the probe was already finalized — prevents double-counting when
        on_failure() and cancel_probe() both fire for the same probe.
        """
        if self._state != CircuitState.HALF_OPEN:
            return
        total_finalized = self._half_open_successes + self._half_open_failures
        if total_finalized >= self._half_open_active:
            return  # All admitted probes already reported — this is a duplicate
        self._half_open_failures += 1
        self._evaluate_half_open()

    def reset(self) -> None:
        """Manually reset the breaker to CLOSED."""
        self._state = CircuitState.CLOSED
        self._consecutive_failures = 0
        self._opened_at = 0.0
        self._half_open_active = 0
        self._half_open_successes = 0
        self._half_open_failures = 0
        self.cooldown_seconds = self._base_cooldown


# ---------------------------------------------------------------------------
# Kill Switch
# ---------------------------------------------------------------------------


class KillSwitch:
    """File-based kill switch.

    Checks for the existence of ``{stateDir}/llmix2/killswitch``.
    If the file exists, all calls are blocked.
    """

    def __init__(self, state_dir: Path | None = None) -> None:
        base = state_dir or _resolve_state_dir()
        self._path = base / _KILLSWITCH_FILENAME

    @property
    def path(self) -> Path:
        return self._path

    def check(self) -> None:
        """Raise KillSwitchActiveError if the kill switch file exists."""
        try:
            self._path.stat()
            raise KillSwitchActiveError(self._path)
        except FileNotFoundError:
            return

    def is_active(self) -> bool:
        """Return True if the kill switch file exists."""
        try:
            self._path.stat()
            return True
        except FileNotFoundError:
            return False


# ---------------------------------------------------------------------------
# Singleflight
# ---------------------------------------------------------------------------


class Singleflight:
    """Deduplicates concurrent identical async calls.

    Uses SHA-256 to key in-flight requests. Concurrent callers with the same
    key share a single future; the result (or error) is propagated to all
    waiters.
    """

    def __init__(self) -> None:
        self._in_flight: dict[str, asyncio.Future[Any]] = {}

    @staticmethod
    def make_key(data: str) -> str:
        """Generate a SHA-256 hex key from a string."""
        return hashlib.sha256(data.encode()).hexdigest()

    async def do(self, key: str, fn: Callable[[], Awaitable[T]]) -> T:
        """Execute fn or join an existing in-flight call for the same key.

        Args:
            key: Deduplication key (use ``make_key`` for raw strings).
            fn: Async callable to execute if no in-flight request exists.

        Returns:
            The result of fn.

        Raises:
            Any exception raised by fn is propagated to all waiters.
        """
        if key in self._in_flight:
            # Shield so a cancelled follower doesn't poison the shared Future
            return await asyncio.shield(self._in_flight[key])

        loop = asyncio.get_running_loop()
        future: asyncio.Future[T] = loop.create_future()
        # Prevent "Future exception was never retrieved" when no second caller awaits
        future.add_done_callback(lambda f: f.exception() if f.done() and not f.cancelled() else None)
        self._in_flight[key] = future

        try:
            result = await fn()
            future.set_result(result)
            return result
        except BaseException as exc:
            future.set_exception(exc)
            raise
        finally:
            self._in_flight.pop(key, None)

    @property
    def in_flight_count(self) -> int:
        return len(self._in_flight)


# ---------------------------------------------------------------------------
# Retry with Exponential Backoff + Jitter
# ---------------------------------------------------------------------------


def calculate_delay(
    attempt: int, *, base_ms: int = _DEFAULT_BASE_DELAY_MS, max_delay_ms: int = _DEFAULT_MAX_DELAY_MS, jitter_ms: int = _DEFAULT_JITTER_MS
) -> int:
    """Calculate retry delay in milliseconds.

    Formula: min(2^attempt * base_ms, max_delay_ms) + random(0, jitter_ms)
    """
    exponential = min((2**attempt) * base_ms, max_delay_ms)
    jitter = random.randint(0, jitter_ms)
    return exponential + jitter


def parse_retry_after(header_value: str | None, *, max_ms: int = _DEFAULT_MAX_RETRY_AFTER_MS) -> int | None:
    """Parse Retry-After header value to milliseconds.

    Supports integer seconds and HTTP-date (RFC 7231) formats.
    Returns None if unparseable. Caps at max_ms.
    """
    if header_value is None:
        return None
    try:
        seconds = int(header_value)
        if seconds < 0:
            return None
        return min(seconds * 1000, max_ms)
    except (ValueError, TypeError):
        pass

    # Fallback: try HTTP-date format (RFC 7231 §7.1.1.1)
    from email.utils import parsedate_to_datetime

    try:
        dt = parsedate_to_datetime(header_value)
        delta_ms = int((dt.timestamp() - time.time()) * 1000)
        if delta_ms > 0:
            return min(delta_ms, max_ms)
    except (ValueError, TypeError):
        pass

    return None


def is_retryable(status_code: int) -> bool:
    """Return True if the status code is retryable (408, 429, or 5xx)."""
    return _is_retryable_status(status_code)


class RetryPolicy:
    """Retry with exponential backoff and jitter.

    Can be used as an async context or called directly.
    """

    def __init__(
        self,
        max_retries: int = 3,
        base_ms: int = _DEFAULT_BASE_DELAY_MS,
        max_delay_ms: int = _DEFAULT_MAX_DELAY_MS,
        jitter_ms: int = _DEFAULT_JITTER_MS,
        max_retry_after_ms: int = _DEFAULT_MAX_RETRY_AFTER_MS,
    ) -> None:
        self.max_retries = max_retries
        self.base_ms = base_ms
        self.max_delay_ms = max_delay_ms
        self.jitter_ms = jitter_ms
        self.max_retry_after_ms = max_retry_after_ms

    def get_delay_ms(self, attempt: int, retry_after_header: str | None = None) -> int:
        """Get the delay for a given attempt in milliseconds.

        If a Retry-After header is provided and valid, it takes precedence
        (capped at max_retry_after_ms).
        """
        retry_after = parse_retry_after(retry_after_header, max_ms=self.max_retry_after_ms)
        if retry_after is not None:
            return retry_after

        return calculate_delay(attempt, base_ms=self.base_ms, max_delay_ms=self.max_delay_ms, jitter_ms=self.jitter_ms)

    async def execute(self, fn: Callable[[], Awaitable[T]], *, is_retryable_fn: Callable[[BaseException], bool] | None = None) -> T:
        """Execute fn with retries.

        Args:
            fn: Async callable to retry.
            is_retryable_fn: Optional predicate to determine retryability.
                Defaults to always-retryable for any exception.
        """
        last_exception: Exception | None = None
        for attempt in range(self.max_retries + 1):
            try:
                return await fn()
            except (asyncio.CancelledError, KeyboardInterrupt, SystemExit):
                raise
            except Exception as exc:
                last_exception = exc
                if attempt >= self.max_retries:
                    raise
                if is_retryable_fn and not is_retryable_fn(exc):
                    raise
                retry_after = getattr(exc, "headers", {})
                retry_after_val = retry_after.get("retry-after") if isinstance(retry_after, dict) else None
                delay_ms = self.get_delay_ms(attempt, retry_after_val)
                await asyncio.sleep(delay_ms / 1000.0)

        # Should never reach here, but satisfy type checker
        assert last_exception is not None
        raise last_exception


# ---------------------------------------------------------------------------
# Cross-Process File Lock (opt-in via LLM_GLOBAL_CONCURRENCY env var)
# ---------------------------------------------------------------------------


class FileLock:
    """Cross-process file lock using fcntl.flock().

    Only active when ``LLM_GLOBAL_CONCURRENCY`` env var is set.
    When not set, acquire/release are no-ops.
    """

    def __init__(self, lock_path: Path | None = None) -> None:
        concurrency = os.environ.get("LLM_GLOBAL_CONCURRENCY")
        self._enabled = concurrency is not None and concurrency.strip() != ""

        if self._enabled:
            base = _resolve_state_dir()
            self._lock_path = lock_path or (base / "llmix.lock")
            self._fd: int | None = None
        else:
            self._lock_path = None
            self._fd = None

    @property
    def enabled(self) -> bool:
        return self._enabled

    def acquire(self) -> None:
        """Acquire the file lock (blocking). No-op if not enabled or on Windows."""
        if not self._enabled or self._lock_path is None:
            return
        if _fcntl is None:
            return
        self._lock_path.parent.mkdir(parents=True, exist_ok=True)
        fd = os.open(str(self._lock_path), os.O_CREAT | os.O_RDWR, 0o600)
        try:
            _fcntl.flock(fd, _fcntl.LOCK_EX)
        except Exception:
            os.close(fd)
            raise
        self._fd = fd

    def release(self) -> None:
        """Release the file lock. No-op if not enabled or not acquired."""
        if self._fd is not None:
            if _fcntl is not None:
                _fcntl.flock(self._fd, _fcntl.LOCK_UN)
            os.close(self._fd)
            self._fd = None

    def __enter__(self) -> FileLock:
        self.acquire()
        return self

    def __exit__(self, *_: Any) -> None:
        self.release()


# ---------------------------------------------------------------------------
# Exports
# ---------------------------------------------------------------------------

__all__ = [
    "CircuitBreaker",
    "CircuitOpenError",
    "CircuitState",
    "FileLock",
    "KillSwitch",
    "KillSwitchActiveError",
    "RetryPolicy",
    "Singleflight",
    "calculate_delay",
    "is_retryable",
    "parse_retry_after",
]
